retis/src/module/skb/bpf/include/if_vlan.h

```diff
@@ -28,21 +32,51 @@ struct skb_vlan_event {
 	u8 acceleration;
 } __binding;
 
+/**
+ * vlan_tag_present tries to determine if a VLAN tag is present. There have been various changes over the years in
+ * the kernel: The kernel uses skb_vlan_tag_present which in more recent versions either relies on vlan_present or on
+ * vlan_all.
+ * We use CO-RE functionality to probe if either field is present and return false otherwise.
+ * See commit 354259fa73e2aac92ae5e19522adb69a92c15b49.
+ *
+ * @param skb
+ *
+ * Returns true if a VLAN tag is present in vlan_present or vlan_all
+ */
 static __always_inline bool vlan_tag_present(const struct sk_buff *skb)
 {
-       struct sk_buff___6_1_0 *skb_61 = (struct sk_buff___6_1_0 *)skb;
+	struct sk_buff___6_1_0 *skb_61 = (struct sk_buff___6_1_0 *)skb;
+
+	// Older kernerls, e.g. RHEL 9.
+	if (bpf_core_field_exists(skb_61->vlan_present))
+		return BPF_CORE_READ_BITFIELD_PROBED(skb_61, vlan_present);
+
+	// New kernels, e.g. Fedora 40 and later.
+	if (bpf_core_field_exists(skb->vlan_all))
+		return BPF_CORE_READ(skb, vlan_all);
 
-       if (bpf_core_field_exists(skb_61->vlan_present))
-               return BPF_CORE_READ_BITFIELD_PROBED(skb_61, vlan_present);
+	return false;
+}
 
-       return BPF_CORE_READ(skb, vlan_all);
+/**
+ * vlan_tag_present_old implements logic from before commit 0c4b2d370514cb4f3454dd3b18f031d2651fab73 to determine if a
+ * VLAN tag is present. In old kernel versions, __vlan_hwaccel_put_tag set skb->vlan_tci = VLAN_TAG_PRESENT | vlan_tci.
+ * Therefore, old versions of the kernel use the DEI / CFI bit indicate / detect presence of the VLAN tag.
+ *
+ * @param skb
+ *
+ * Returns true if a VLAN tag is present
+ */
+static __always_inline bool vlan_tag_present_old(const struct sk_buff *skb) {
+	return (BPF_CORE_READ(skb, vlan_tci) & VLAN_TAG_PRESENT);
 }
 
 /**
- * __vlan_hwaccel_get_tag - get the VLAN ID that is in @skb->cb[]
- * The kernel uses skb_vlan_tag_present which either relies on vlan_present or
- * on vlan_all depending on the kernel version (see commit 354259fa73e2aac92ae5e19522adb69a92c15b49).
- * We use CO-RE functionality to probe either field in vlan_tag_present.
+ * __vlan_hwaccel_get_tag - get the VLAN TCI that is in @skb->cb[]. In newer versions of the kernel, report the vlan_tci
+ * as is. Versions of the kernel before commit 0c4b2d370514cb4f3454dd3b18f031d2651fab73 always set the DEI / CFI bit to
+ * on. As it is more common to have this bit off, and in order to not report a false positive, set the bit to 0 for 
```

Comment:
nit: extra space at EOL.

Also in general can you wrap the code + comments to 80 chars (when possible)? Thanks!

Reply:
done